pulsar-broker/src/main/java/org/apache/pulsar/compaction/PulsarTopicCompactionService.java

```diff
@@ -106,6 +111,78 @@ public CompletableFuture<Position> getLastCompactedPosition() {
         return CompletableFuture.completedFuture(compactedTopic.getCompactionHorizon().orElse(null));
     }
 
+    @Override
+    public CompletableFuture<Entry> findEntryByPublishTime(long publishTime) {
+        final Predicate<Entry> predicate = entry -> {
+            try {
+                return Commands.parseMessageMetadata(entry.getDataBuffer()).getPublishTime() >= publishTime;
+            } finally {
+                entry.release();
+            }
+        };
+        return findFirstMatchEntry(predicate);
+    }
+
+    @Override
+    public CompletableFuture<Entry> findEntryByEntryIndex(long entryIndex) {
+        final Predicate<Entry> predicate = entry -> {
+            try {
+                BrokerEntryMetadata brokerEntryMetadata =
+                        Commands.parseBrokerEntryMetadataIfExist(entry.getDataBuffer());
+                if (brokerEntryMetadata == null || !brokerEntryMetadata.hasIndex()) {
+                    return false;
+                }
+                return brokerEntryMetadata.getIndex() >= entryIndex;
+            } finally {
+                entry.release();
+            }
+        };
+        return findFirstMatchEntry(predicate);
+    }
+
+    private CompletableFuture<Entry> findFirstMatchEntry(final Predicate<Entry> predicate) {
+        var compactedTopicContextFuture = compactedTopic.getCompactedTopicContextFuture();
+
+        if (compactedTopicContextFuture == null) {
+            return CompletableFuture.completedFuture(null);
+        }
+        return compactedTopicContextFuture.thenCompose(compactedTopicContext -> {
+            LedgerHandle lh = compactedTopicContext.getLedger();
+            CompletableFuture<Long> promise = new CompletableFuture<>();
+            findFirstMatchIndexLoop(predicate, 0L, lh.getLastAddConfirmed(), promise, null, lh);
+            return promise.thenCompose(index -> {
+                if (index == null) {
+                    return CompletableFuture.completedFuture(null);
+                }
+                return readEntries(lh, index, index).thenApply(entries -> entries.get(0));
+            });
+        });
+    }
+
+    private static void findFirstMatchIndexLoop(final Predicate<Entry> predicate,
+                                           final long start, final long end,
+                                           final CompletableFuture<Long> promise,
+                                           final Long lastMatchIndex,
+                                           final LedgerHandle lh) {
+        if (start > end) {
+            promise.complete(lastMatchIndex);
+            return;
+        }
+
+        long mid = (start + end) / 2;
+        CompactedTopicImpl.readEntries(lh, mid, mid).thenAccept(entries -> {
+            Entry entry = entries.get(0);
+            if (predicate.test(entry)) {
+                findFirstMatchIndexLoop(predicate, start, mid - 1, promise, mid, lh);
```

Comment:
I have some concerns here:

1. Is it possible to get `StackOverflow` here?
2. In the case of a large amount of data, will it take up the key threads of BK?

Reply:
1. I don't think so, since `readEntries` will switch threads.
2. Yes, It will take up the callback thread of BK, but it shouldn't be too long and other tasks still have the opportunity to use this thread.

Reply:
Got it. thanks!